meteor/server/api/deviceTriggers/TagsService.ts

```diff
@@ -0,0 +1,168 @@
+			activationId: {
+				$exists: true,
+			},
+		})
+		if (!rundownPlaylist) {
+			return false
+		}
+
+		const previousPartInstanceId = rundownPlaylist?.previousPartInfo?.partInstanceId
+		const currentPartInstanceId = rundownPlaylist?.currentPartInfo?.partInstanceId
+		const nextPartInstanceId = rundownPlaylist?.nextPartInfo?.partInstanceId
+
+		const showStyleBase = cache.ShowStyleBases.findOne(showStyleBaseId)
+
+		if (!showStyleBase) return false
+
+		const resolvedSourceLayers = applyAndValidateOverrides(showStyleBase.sourceLayersWithOverrides).obj
+
+		const inPreviousPartInstance = previousPartInstanceId
+			? this.processAndPrunePieceInstanceTimings(
+					cache.PartInstances.findOne(previousPartInstanceId)?.timings,
+					cache.PieceInstances.find({ partInstanceId: previousPartInstanceId }).fetch(),
+					resolvedSourceLayers
+			  )
+			: []
+		const inCurrentPartInstance = currentPartInstanceId
+			? this.processAndPrunePieceInstanceTimings(
+					cache.PartInstances.findOne(currentPartInstanceId)?.timings,
+					cache.PieceInstances.find({ partInstanceId: currentPartInstanceId }).fetch(),
+					resolvedSourceLayers
+			  )
+			: []
+		const inNextPartInstance = nextPartInstanceId
+			? this.processAndPrunePieceInstanceTimings(
+					undefined,
+					cache.PieceInstances.find({ partInstanceId: nextPartInstanceId }).fetch(),
+					resolvedSourceLayers
+			  )
+			: []
+
+		const activePieceInstances = [...inPreviousPartInstance, ...inCurrentPartInstance].filter((pieceInstance) =>
+			this.isPieceInstanceActive(pieceInstance, previousPartInstanceId, currentPartInstanceId)
+		)
+
+		const activePieceInstancesTags = new Set<string>()
+		activePieceInstances.forEach((pieceInstance) => {
+			pieceInstance.piece.tags?.forEach((tag) => {
+				activePieceInstancesTags.add(tag)
+			})
+		})
+
+		const nextPieceInstancesTags = new Set<string>()
+		inNextPartInstance.forEach((pieceInstance) => {
+			pieceInstance.piece.tags?.forEach((tag) => {
+				nextPieceInstancesTags.add(tag)
+			})
+		})
+
+		const shouldUpdateTriggers = this.shouldUpdateTriggers(activePieceInstancesTags, nextPieceInstancesTags)
+
+		this.onAirPiecesTags = activePieceInstancesTags
+		this.nextPiecesTags = nextPieceInstancesTags
+
+		return shouldUpdateTriggers
+	}
+
+	private shouldUpdateTriggers(activePieceInstancesTags: Set<string>, nextPieceInstancesTags: Set<string>) {
+		return (
+			(!this.areSetsEqual(this.onAirPiecesTags, activePieceInstancesTags) ||
+				!this.areSetsEqual(this.nextPiecesTags, nextPieceInstancesTags)) &&
+			(this.doSetsIntersect(activePieceInstancesTags, this.tagsObservedByTriggers) ||
+				this.doSetsIntersect(nextPieceInstancesTags, this.tagsObservedByTriggers) ||
+				this.doSetsIntersect(this.onAirPiecesTags, this.tagsObservedByTriggers) ||
+				this.doSetsIntersect(this.nextPiecesTags, this.tagsObservedByTriggers))
+		)
+	}
+
+	protected areSetsEqual<T>(a: Set<T>, b: Set<T>): boolean {
```

Comment:
Why are these methods protected? I'm also wondering if they shouldn't live in some lib file somewhere, considering they are super-generic and don't reference the members of this class at all?